### PR TITLE
[CIS-2179] Add support for bypassing the messages and replies data source

### PR DESCRIPTION
**Pull request by nuno-vieira**
### 🔗 Issue Links
CIS-2179
https://github.com/GetStream/stream-chat-swift/issues/2305

### 🎯 Goal
Add support for bypassing the messages and replies data source.

### 🛠 Implementation
Provide a `Components.messagesBypass` and `Components.repliesBypass` to be able to override the message list data source. It is also possible to change directly inside the `ChatThreadVC` and `ChatChannelVC`.

TODO:
- [ ] Add DemoApp example to test this
- [ ] Finish test coverage
- [ ] Refactor and reuse some logic between Thread and Channel

Limitations:
- The pagination number won't be respected. If the pagination is 25 by 25, for example, since the filter happens after fetching from the DB. If the bypass removes 2 messages, it will be loading 23, and not 25. Either way, the same thing happens already when hiding deleted messages on the client side.

### 🧪 Manual Testing Notes
TODO: Add an example in the DemoApp

### ☑️ Contributor Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] This change follows zero ⚠️ policy (required)
- [x] This change should be manually QAed
- [ ] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (docusaurus, tutorial, CMS)

### CODE REVIEW

**Review by martinmitrevski — COMMENTED** [review]
Looks good 👍 
My two cents on: "Another option would be to provide a ([ChatMessage] -> [ChatMessages]) closure instead of ([ChatMessage] -> Bool) to allow for even more flexibility" - I would go with this option, since it allows more flexibility. Apart from adding more messages, you could also check dependencies between messages.
Additionally, we would need changelog update and docusaurus docs.

**Comment by polqf** [comment]
Not really a fan of having filtering logic duplicated in each file. Is there any way to abstract it?

**Comment by nuno-vieira** [comment]
> Not really a fan of having filtering logic duplicated in each file. Is there any way to abstract it?

What do you mean by each file? 🤔

File: Sources/StreamChatUI/ChatThread/ChatThreadVC.swift [hunk 1]
```diff
@@ -321,8 +331,16 @@ open class ChatThreadVC: _ViewController,
...
-        messageListVC.setNewMessagesSnapshot(messages)
+        let newMessages: [ChatMessage]
+        if let repliesBypass = repliesBypass {
+            newMessages = repliesBypass(messages)
+        } else {
+            newMessages = messages
+        }
+
+        messageListVC.setNewMessagesSnapshot(newMessages)
         messageListVC.updateMessages(with: changes)
```
Comment (nuno-vieira) [inline Sources/StreamChatUI/ChatThread/ChatThreadVC.swift:344]: This is a problem. We are not filtering the `changes` and use it for some logic inside the `messageListVC` 🤔

Comment (nuno-vieira) [reply Sources/StreamChatUI/ChatThread/ChatThreadVC.swift]: This is now fixed 🙂

File: Sources/StreamChatUI/ChatThread/ChatThreadVC.swift [hunk 2]
```diff
@@ -320,10 +330,22 @@ open class ChatThreadVC: _ViewController,
...
+        let newMessages: [ChatMessage]
+        let newChanges: [ListChange<ChatMessage>]
+        if let repliesBypass = repliesBypass {
+            newMessages = repliesBypass(messages)
+            let messageIds = Set(newMessages.map(\.id))
+            newChanges = changes.filter { messageIds.contains($0.item.id) }
+        } else {
+            newMessages = messages
+            newChanges = changes
+        }
```
Comment (nuno-vieira) [inline Sources/StreamChatUI/ChatThread/ChatThreadVC.swift:344]: This can be easily extracted and reused between Channel and Thread

File: Sources/StreamChatUI/ChatThread/ChatThreadVC.swift [hunk 3]
```diff
@@ -150,11 +155,16 @@ open class ChatThreadVC: _ViewController,
 
     // MARK: - ChatMessageListVCDataSource
 
+    // The replies data source.
     public var messages: [ChatMessage] {
         get {
             replies
```
Comment (ipavlidakis) [inline Sources/StreamChatUI/ChatThread/ChatThreadVC.swift:161]: I have to admit that seeing returning replies for messages is a little bit confusing (i know this is not part of your changes just mentioning it)

File: Tests/StreamChatUITests/SnapshotTests/ChatChannel/ChatChannelVC_Tests.swift [hunk 4]
```diff
@@ -549,6 +549,24 @@ final class ChatChannelVC_Tests: XCTestCase {
...
+
+        /// Regular case
+        vc.messagesBypass = { messages in messages.filter(\.isBounced) }
+        vc.messages = [.mock(isBounced: false), .mock(isBounced: true), .mock(isBounced: true)]
+        XCTAssertEqual(vc.messages.count, 2)
+    }
+
+    func test_messagesBypass_whenDidChangeMessages_shouldOverrideNewSnapshot_shouldOverrideChanges() {
+        XCTFail()
+    }
```
Comment (ipavlidakis) [inline Tests/StreamChatUITests/SnapshotTests/ChatChannel/ChatChannelVC_Tests.swift:569]: Do you mind adding a test with the messageBypass being nil just to ensure that we are not affecting the default behaviour?

File: Sources/StreamChatUI/Components.swift [hunk 5]
```diff
@@ -100,6 +100,12 @@ public struct Components {
     /// Used in both the Channel and Thread view controllers.
     @available(iOSApplicationExtension, unavailable)
     public var messageListVC: ChatMessageListVC.Type = ChatMessageListVC.self
+
+    /// A closure to bypass the message list channel data source.
+    public var messagesBypass: (([ChatMessage]) -> [ChatMessage])?
```
Comment (ipavlidakis) [inline Sources/StreamChatUI/Components.swift:105]: One thing that I would like to point out is that blocks are more difficult to debug and provide useful information in stacktraces (after a crash). I would be more confident with a new protocol(maybe MessageFiltering) & object(provided from clients) pair.
